_posts/2023-03-15-Automate-Software-Workflows-with-Github-Actions.md

```diff
@@ -0,0 +1,148 @@
+
+## Nested Workflows with Github Actions
+
+<pre class="mermaid">
+flowchart LR
+
+  subgraph action ["Nested Workflow (Dagger, etc)"]
+    direction LR
+    actions
+    start2((start)) --> actions
+    actions --> en2((end))
+    en2((end))
+  end
+  subgraph workflow2 [Local Environment Run]
+    direction LR
+    run2[run workflow]
+    en3((end))
+    start3((event\ntrigger))
+  end
+  subgraph workflow [Github Actions Workflow Run]
+    direction LR
+    start((event\ntrigger))
+    run[run workflow]
+    en((end))
+  end
+  
+  start --> run
+  start3 --> run2
+  action -.-> run
+  run --> en
+  run2 --> en3
+  action -.-> run2
+  style start fill:#6EE7B7
+  style start2 fill:#D1FAE5
+  style start3 fill:#6EE7B7
+  style en fill:#FCA5A5
+  style en2 fill:#FFE4E6
+  style en3 fill:#FCA5A5
+</pre>
+
+_A diagram showing how Github Actions may leverage nested workflows with tools like Dagger._
+{:.center}
+
+There are times when Github Actions may be too constricting or Act may not accurately simulate workflows. We also might seek to ["write once, run anywhere" (WORA)](https://en.wikipedia.org/wiki/Write_once,_run_anywhere) to enable flexible development on many environments. One workaround to this challenge is to use nested workflows which are compatible with local environments and Github Actions environments. [Dagger](https://docs.dagger.io/) is one tool which enables programmatically specifying and using workflows this way. Using Dagger allows you to trigger workflows on your local machine or Github Actions with the same underlying engine, meaning there are fewer inconsistencies or guesswork for developers (see here for [an explanation of how Dagger works](https://docs.dagger.io/cli#how-does-it-work)).
+
+> __Github Actions with Nested Workflow Example__
+> Reference this example for a brief demonstration of how Github Actions and Dagger may be used together.
+>
+> - [4.run-matrixed-pytest-dagger.yml](https://github.com/CU-DBMI/demo-github-actions/blob/main/.github/workflows/4.run-matrixed-pytest-dagger.yml): demonstrates how to run matrixed Python versions for confirming passing pytest tests using Github Actions and Dagger together. A [Github Actions matrix strategy](https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs) is used to span concurrent work while retaining the reproducibility from Dagger task specification.
+
+There are also other alternatives to Dagger you may want to consider based on your usecase, preference, or interest:
+
+- [Earthly](https://github.com/earthly/earthly): similar to Dagger and uses "earthfiles" as a specification.
+- [Docker (Dockerfiles)](https://docs.docker.com/engine/reference/builder/): it's worth noting that Dockerfiles and other Docker technologies may enable decoupled workflow development in a similar way to Dagger and Earthly. One disadvantage of this approach is how coupled your stack may become to Docker (including potential licensing expenses).
```

Comment:
I don't know; I see the same coupling issue you're describing occurring Dagger, Earthly, or really any other CI/CD system, and Docker is at least closer to the source of running containers directly rather than through an intermediate layer. You could feasibly replace your Docker image building and execution system with fully free and open-source alternatives, but you're kind of at the mercy of Dagger for how things work under the hood. AFAIK Dagger runs jobs in containers anyway, and the most accessible container runtime is Docker, so if you go with the path of least resistance for setting up Dagger you're still subject to the licensing restrictions you mentioned. If you switch to an alternative open-source container runtime to use with Dagger then you could have used that with bare containers anyway...

This is all to say that it's a complicated issue that deserves further discussion; I think people should be encouraged to learn about how images are built and containers are run versus relying on Dagger or some other system to abstract it away, since they use containers internally anyway. I guess my suggestion would be to drop the last sentence ("One disadvantage..."), or if you're up to it expand on how Docker is part of a larger ecosystem of tools that all comply with the Open Container Initiative standards, and that, while there are licensing issues with Docker (outside of research and academia), these licensing issues shouldn't preclude one from considering directly using container technology.

Reply:
Thanks, great thoughts! I think you're right, I was unfair here and was intending to warn about potential costs involved with Docker specifically. Dagger leverages [Buildkit](https://github.com/moby/buildkit) as the basis of running containers instead of Docker-based solutions, so it seems more flexible about the cost considerations. For example, it's possible to execute Dagger through Podman, and I could see why someone would want to do this for technical, business, or cultural reasons.

I felt less certain about adding Docker here in general as it's used differently and might need more work than Dagger or Earthly to setup in this way. This seems like some of the reason these projects were created - as an evolution of containerized workflow management (alongside Github Actions). From a topic standpoint, Docker feels like a related but separate space when compared with Github Actions, Dagger, and Earthly (workflow systems leveraging containers). Maybe a general statement about what you mentioned on containers would be apt; that these are all virtualized workflow systems which wrap around the idea of reproducibility through containers (whether it be Docker, Podman, Buildkit, etc?

Reply:
> Maybe a general statement about what you mentioned on containers would be apt; that these are all virtualized workflow systems which wrap around the idea of reproducibility through containers (whether it be Docker, Podman, Buildkit, etc?

I like that; it gives enough perspective on where these workflow sit without getting into the nitty gritty of containerization. On that note, I agree that talking about Docker and containerization in general is probably outside the scope of this post. I wasn't aware that Dagger used BuildKit by default; that's pretty cool.

Reply:
I replaced the Docker mention in this block with a generalized mention of containers alongside Earthly, Dagger, and Github Actions. Thanks again for the discussion on this topic!